src/Dfc.CourseDirectory.Functions/DeleteDuplicateProviderRecords.cs

```diff
@@ -0,0 +1,198 @@
+    return
+end
+
+
+-- Consolidate the ProviderType for all records for a given UKPRN, ensuring we don't remove ProviderType(s)
+
+update #DuplicateProviders
+set ProviderType = ProviderType | 1
+from #DuplicateProviders x
+join (select Ukprn from Pttcd.Providers where ProviderType & 1 <> 0 group by Ukprn) p on p.Ukprn = x.Ukprn
+
+update #DuplicateProviders
+set ProviderType = ProviderType | 2
+from #DuplicateProviders x
+join (select Ukprn from Pttcd.Providers where ProviderType & 2 <> 0 group by Ukprn) p on p.Ukprn = x.Ukprn
+
+
+-- Pick the provider records to remove (keeping the most recently updated for a given UKPRN)
+;
+with DuplicateProviders as (
+	select ProviderId, Ukprn, ROW_NUMBER() over (partition by Ukprn order by Ukprn, UpdatedOn desc) RowNum
+	from Pttcd.Providers
+	where Ukprn in (
+		select Ukprn from #DuplicateProviders
+	)
+)
+select ProviderId, Ukprn into #DeletingProviderIds from DuplicateProviders
+where RowNum > 1
+
+
+-- Find the remaining provider IDs where their ProviderType has been changed as a result of the consolidation above
+select p.ProviderId, p.Ukprn, p.ProviderType OldProviderType, y.ProviderType NewProviderType
+from Pttcd.Providers p
+left join #DeletingProviderIds x on x.ProviderId = p.ProviderId
+join #DuplicateProviders y on p.Ukprn = y.Ukprn
+where x.ProviderId is null and y.ProviderType <> p.ProviderType
+
+
+if @DryRun = 0
+begin
+	-- Remove provider records
+	delete from Pttcd.ProviderContacts where ProviderId in (select ProviderId from #DeletingProviderIds)
+	delete from Pttcd.Providers where ProviderId in (select ProviderId from #DeletingProviderIds)
+
+	-- Assign consolidate ProviderType
+	update Pttcd.Providers
+	set ProviderType = x.ProviderType
+	from Pttcd.Providers p
+	join #DuplicateProviders x on p.Ukprn = x.Ukprn
+end
+
+
+-- Return the provider records we've deleted
+select d.ProviderId, p.Ukprn from #DeletingProviderIds d
+join #DuplicateProviders p ON d.Ukprn = p.Ukprn
+
+
+drop table #DuplicateProviders
+drop table #DeletingProviderIds";
+
+            IList<UpdatedProviderInfo> updatedProviders;
+            IList<DeletedProviderInfo> deletedProviders;
+
+            using (var reader = await sqlConnection.QueryMultipleAsync(sql, new { DryRun = dryRun }))
+            {
+                updatedProviders = (await reader.ReadAsync<UpdatedProviderInfo>()).AsList();
+                deletedProviders = (await reader.ReadAsync<DeletedProviderInfo>()).AsList();
+            }
+
+            if (!dryRun)
+            {
+                foreach (var record in updatedProviders)
+                {
+                    await _cosmosDbQueryDispatcher.ExecuteQuery(new UpdateProviderType()
+                    {
+                        ProviderId = record.ProviderId,
+                        ProviderType = record.NewProviderType
+                    });
+                }
+
+                foreach (var record in deletedProviders)
```

Comment:
what happens if the sql fails having deleted some, would we then have to work harder to fixup cosmos?

Reply:
We could re-run the full sync to get records back into SQL then re-run. Given that we don't have distributed transactions it's never going to be perfect.